File: Graph/csharp/SemestersRequired.cs
using System.Collections.Generic;

namespace GraphSolutions;

public static class SemestersRequired
{
    public static int Solve(int numCourses, IEnumerable<(int, int)> prereqs)
    {
        var graph = new Dictionary<int, List<int>>();
        for (var i = 0; i < numCourses; i++)
        {
            graph[i] = new List<int>();
        }

        foreach (var (a, b) in prereqs)
        {
            graph[a].Add(b);
        }

        var distance = new Dictionary<int, int>();
        foreach (var (node, neighbors) in graph)
        {
            if (neighbors.Count == 0)
            {
                distance[node] = 1;
            }
        }

        var maxSemester = 0;
        for (var i = 0; i < numCourses; i++)
        {
            var value = Traverse(graph, i, distance);
            if (value > maxSemester)
            {
                maxSemester = value;
            }
        }

        return maxSemester;
    }

    private static int Traverse(Dictionary<int, List<int>> graph, int node, Dictionary<int, int> distance)
    {
        if (distance.TryGetValue(node, out var cached))
        {
            return cached;
        }

        var maxChild = 0;
        foreach (var neighbor in graph[node])
        {
            var candidate = Traverse(graph, neighbor, distance);
            if (candidate > maxChild)
            {
                maxChild = candidate;
            }
        }

        distance[node] = maxChild + 1;
        return distance[node];
    }
}

File: Graph/ts/semestersRequired.ts
export function semestersRequired(numCourses: number, prereqs: Array<[number, number]>): number {
  const graph: Record<number, number[]> = {};
  for (let i = 0; i < numCourses; i += 1) graph[i] = [];
  for (const [a, b] of prereqs) {
    graph[a].push(b);
  }

  const distance: Record<number, number> = {};
  for (let i = 0; i < numCourses; i += 1) {
    if (graph[i].length === 0) distance[i] = 1;
  }

  let maxSemester = 0;
  for (let i = 0; i < numCourses; i += 1) {
    const value = traverse(graph, i, distance);
    if (value > maxSemester) maxSemester = value;
  }

  return maxSemester;
}

function traverse(graph: Record<number, number[]>, node: number, distance: Record<number, number>): number {
  if (distance[node] !== undefined) return distance[node];

  let longest = 0;
  for (const neighbor of graph[node]) {
    const candidate = traverse(graph, neighbor, distance);
    if (candidate > longest) longest = candidate;
  }

  distance[node] = longest + 1;
  return distance[node];
}
